core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,324 @@
+    val start = transid.started(
+      this,
+      LoggingMarkers.INVOKER_KUBECTL_CMD(args.head),
+      s"running ${cmd.mkString(" ")} (timeout: $timeout)",
+      logLevel = InfoLevel)
+    executeProcess(cmd, timeout).andThen {
+      case Success(_) => transid.finished(this, start)
+      case Failure(t) => transid.failed(this, start, t.getMessage, ErrorLevel)
+    }
+  }
+}
+
+object KubernetesClient {
+
+  //%Y-%m-%dT%H:%M:%S.%N%z
+  val K8SDateTimeFormat = new DateTimeFormatterBuilder()
+    .parseCaseInsensitive()
+    .appendPattern("u-MM-dd")
+    .appendLiteral('T')
+    .appendPattern("HH:mm:ss[.n]")
+    .appendLiteral('Z')
+    .toFormatter()
+
+}
+
+trait KubernetesApi {
+  def run(name: String, image: String, args: Seq[String] = Seq.empty[String])(
+    implicit transid: TransactionId): Future[ContainerId]
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress]
+
+  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit]
+
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit]
+
+  def logs(containerId: ContainerId, sinceTime: Option[LocalDateTime], waitForSentinel: Boolean = false)(
+    implicit transid: TransactionId): Source[ByteString, Any]
+}
+
+object KubernetesRestLogSourceStage {
+
+  import KubernetesClient.K8SDateTimeFormat
+
+  def constructPath(namespace: String, containerId: String): Path =
+    Path / "api" / "v1" / "namespaces" / namespace / "pods" / containerId / "log"
+
+  def constructQuery(sinceTime: Option[LocalDateTime], waitForSentinel: Boolean): Query = {
+
+    val qB = Map.newBuilder[String, String]
+    qB += "timestamps" → "true"
+    qB ++= sinceTime.map(time => "sinceTime" -> time.format(KubernetesClient.K8SDateTimeFormat))
+
+    Query(qB.result())
+  }
+
+  @tailrec
+  def readLines(src: BufferedSource,
+                lastTimestamp: Option[LocalDateTime],
+                lines: Seq[JsObject] = Seq.empty[JsObject]): (Option[LocalDateTime], Seq[JsObject]) = {
+
+    if (!src.exhausted()) {
+      (for {
+        l <- Option(src.readUtf8Line()) if !l.isEmpty
+        p = l.indexOf(" ")
+        // Kubernetes is ignoring nanoseconds in sinceTime, so we have to filter additionally here
+        ts = l.substring(0, p)
+        tsDate <- Try(LocalDateTime.parse(ts, K8SDateTimeFormat)).toOption if isRelevantLogLine(lastTimestamp, tsDate)
+        msg = l.substring(p + 1)
+        stream = "stdout" // TODO - when we can distinguish stderr: https://github.com/kubernetes/kubernetes/issues/28167
+      } yield {
+        tsDate -> JsObject("log" -> msg.toJson, "stream" -> stream.toJson, "time" -> ts.toJson)
+      }) match {
+        case Some((latestTS, js)) =>
+          readLines(src, Some(latestTS), lines :+ js)
+        case None =>
+          // we may have skipped a line for filtering conditions only; keep going
+          readLines(src, lastTimestamp, lines)
+      }
+    } else {
+      src.close()
```

Comment:
Should the caller be responsible of closing the source? Since it isn't opened by `readLines`

Reply:
It struck me at the time as the sanest approach, since the HTTP library will drop lots of exceptions and warnings in  the logs if you don't close the source. But it isn't clean; i'll move it back out to the caller of `readLines`